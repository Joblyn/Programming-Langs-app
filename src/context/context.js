import { createContext } from "react";

export const RouteContext = createContext(null);

export const NameContext = createContext(null);