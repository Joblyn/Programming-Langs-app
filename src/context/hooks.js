export const useAppContext = (value = null) => {
  let route = value;
  return route;
};