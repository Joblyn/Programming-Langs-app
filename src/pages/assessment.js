import React, { useState, useEffect, createElement } from "react";
import NameIcon from "../assets/images/name_icon.svg";
import ProgressLine from "../components/progress/progressline";
import { LargeButton } from "../components/button/button";
import Question from "../components/question/question";
import { useMediaQuery } from "@mui/material";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { getQuestions } from "../utilities/redux/actions/questions";
import Loader from "../components/loader/loader";

export default function Assessment() {
  const { language } = useParams();
  const [count, updateCount] = useState(1);
  const [startEnd, setStartEnd] = useState({ start: 0, end: 5 });
  const [data, setData] = useState([]);

  const [last, setLast] = useState(false);
  const remainder = data.length % 5;
  const quotient = parseInt(data.length / 5);
  const matches = useMediaQuery("(min-width:960px)");
  const dispatch = useDispatch();

  const questions = useSelector((state) => state.questions);
  const [questionsState, setQuestionsState] = useState([]);

  const responseStore = useSelector((state) => state.responses);
  const [responses, setResponses] = useState({});

  const [error, setError] = useState({});

  // dispatch action to fetch questions and store at redux store
  useEffect(() => {
    dispatch(getQuestions());
  }, [language, dispatch]);

  // store all language questions from redux store
  useEffect(() => {
    if (questions.isSuccessful) {
      setData(questions.data);
    }
    if (questions.isError) {
      console.error()
      setError({ type: "fetch error", message: error });
    }
  }, [questions, error]);

  // set current questions and responses sets
  useEffect(() => {
    if (startEnd.end === undefined) {
      setQuestionsState(data.slice(startEnd.start));
    } else {
      setQuestionsState(data.slice(startEnd.start, startEnd.end));
    }
  }, [data, startEnd]);

  // check if current questions set is the last and update state
  useEffect(() => {
    if (remainder === 0) {
      if (count === quotient) {
        setLast(true);
      } else {
        setLast(false);
      }
      return;
    }
  }, [count, remainder, quotient]);

  // update responses state
  useEffect(() => {
    setResponses(function () {
      let obj = {};
      questionsState.forEach((question) => {
        let prop = `question_${question.id}`;
        obj = { ...obj, [prop]: responseStore[prop] || "" };
      });
      return obj;
    });
  }, [questionsState, responseStore]);

  const nextQuestionsSet = () => {
    updateCount((count) => count + 1);
    if (remainder) {
      if (count === quotient) {
        setStartEnd(({ end }) => ({ start: end }));
        setLast(true);
        setError({});
      }
    } else {
      setStartEnd(({ end }) => ({ start: end, end: end + 5 }));
      setError({});
    }
  };

  const previousQuestionSet = () => {
    updateCount((count) => count - 1);
    setStartEnd(({ start, end }) => ({ start: start - 5, end: end - 5 }));
    setError({});
  };

  const createCommponent = () => {
    const getProps = (questions, id) => ({
      className: id === questions.length - 1 ? "pb-[173px] md:pb-[100px] question_container" : "question_container",
      key: "question_" + id,
    });
    return questionsState.map((question, id) => {
      let propName = "question_" + question.id;
      return createElement(
        "div",
        getProps(questionsState, id),
        <Question
          propName={propName}
          id={id}
          defaultValue={responseStore[propName]}
          question={question}
          setResponses={setResponses}
        />
      );
    });
  };

  let incompleteResponses = Object.values(responses).includes("");

  return (
    <div className="w-full relative" id="top">
      <div className="flex items-center my-2 w-full">
        <span className="align-middle mr-3">
          <img src={NameIcon} alt="_logo_" className="w-10 h-10" />
        </span>
        <span className="uppercase text-green text-lg font-bold ">
          {`${language} assessment`}
        </span>
      </div>
      {questionsState.length ? (
        <>
          <ProgressLine count={count} totalcount={quotient} />
          {createCommponent()}
          <div className={`${matches ? "px-6 pb-6" : "assessment_footer"}`}>
            {error.type === "incomplete responses" && (
              <p className="text-[red] text-base px-6">{error.message}</p>
            )}
            <div className="flex gap-3">
              {count !== 1 && (
                <LargeButton
                  type="back"
                  innerText={"Back"}
                  action={previousQuestionSet}
                />
              )}
              <LargeButton
                type="next"
                innerText={last ? "Finish" : "Next"}
                action={
                  incompleteResponses
                    ? () =>
                        setError({
                          type: "incomplete responses",
                          message: "Please fill in all fields",
                        })
                    : last
                    ? null
                    : nextQuestionsSet
                }
              />
            </div>
          </div>
        </>
      ) : (
        <Loader />
      )}
    </div>
  );
}
