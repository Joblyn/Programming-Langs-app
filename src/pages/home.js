import React, { useEffect, useState } from "react";
import CustomCard from "../components/card/card";
import Header from "../components/header/header";
import PageLoader from "../components/pageloader/pageloader";
import Languages from "../fixtures/languages";
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';

export default function Home() {
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setTimeout(() => {
      setLoading(false);
    }, 2000);
  }, []);

  if (loading) {
    return (
      <div className="min-h-screen">
        <PageLoader />
      </div>
    );
  }

  return (
    <div>
      <Header />
      <main className="mt-[6rem] p-2">
        <Box sx={{ flexGrow: 1 }}>
          <Grid
            container
            spacing={{ xs: 2, md: 3 }}
            columns={{ xs: 4, sm: 8, md: 12 }}
          >
            {Languages.map((lang, id) => (
              <Grid item xs={2} sm={4} md={4} key={`lang_${lang}_${id}`}>
                <CustomCard
                  title={lang.title}
                  image={lang.image}
                  alt={lang.alt}
                />
              </Grid>
            ))}
          </Grid>
        </Box>
      </main>
    </div>
  );
}
