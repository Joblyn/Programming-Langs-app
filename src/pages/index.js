export { default as Home } from './home';
export { default as Javascript } from './javascript';
export { default as Python } from './python';
export { default as Java } from "./java";