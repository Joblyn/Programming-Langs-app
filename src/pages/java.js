import React from "react";
import data from "../data/java.json";

export default function Java(){
  return <div>
    
  </div>
}