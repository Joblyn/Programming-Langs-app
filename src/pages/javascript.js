import React from "react";
import { Link } from "react-router-dom";

export default function Javascript() {
  return <div>
    <h1>This is  the javascript page</h1>
    <Link to="/">Home</Link>
  </div>;
}
