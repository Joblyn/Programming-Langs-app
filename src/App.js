import React, { useState, useEffect } from "react";
import "./App.css";
import { Routes, Route } from "react-router-dom";
import * as ROUTES from "./constants/routes";
import { Home, PageNotFound, Assessment, Name } from "./pages";
import Layout from "./containers/layout/layout";
import { AppContext } from "./context/context";
import ReactGA from "react-ga";

// initialize google analytics
let TRACKING_ID;
if (window.location.hostname === "woshipfrontend.interviewblindspots.com") {
  TRACKING_ID = "G-QW0ZP2JZHF";
} else {
  if (window.location.hostname === "damp-shore-08296.herokuapp.com") {
    TRACKING_ID = "G-L3SYMZVBRR";
  } else {
    TRACKING_ID = "G-8WCMDG75N4";
  }
}
ReactGA.initialize(TRACKING_ID);

function App() {
  const [route, setRoute] = useState();
  const [name, setName] = useState();

  useEffect(() => {
    ReactGA.pageview(window.location.pathname + window.location.search);
  }, []);

  return (
    <AppContext.Provider value={{ route, setRoute, setName, name }}>
      <Routes>
        <Route index element={<Home />} />
        <Route path={ROUTES.HOME} element={<Home />} />
        <Route path={ROUTES.ASSESSMENT} exact element={<Layout />}>
          <Route path={ROUTES.NAME} exact element={<Name />} />
          <Route path={ROUTES.LANGUAGE} exact element={<Assessment />} />
        </Route>
        <Route path={ROUTES.NOMATCH} element={<PageNotFound />} />
      </Routes>
    </AppContext.Provider>
  );
}

export default App;
