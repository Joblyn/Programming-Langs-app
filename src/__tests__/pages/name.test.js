import React from "react";
import {
  render,
  cleanup,
  fireEvent,
  getByPlaceholderText,
} from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import "@testing-library/jest-dom";
import { Name } from "../../pages";
import { AppContext } from "../../context/context";
import App from "../../App";
import { act } from "react-dom/test-utils";

// mock useNavigate from React
const mockedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
}));

// test the app at name
describe("<Name />", () => {
  afterEach(cleanup);

  // test name component
  test("renders correctly", async () => {
    const route = "";
    const setName = () => {};
    const history = createMemoryHistory();
    const { container, getByText, getByAltText, getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <AppContext.Provider value={{ route, setName }}>
          <Name />
        </AppContext.Provider>
      </Router>
    );

    // expect to see a logo image
    expect(getByAltText("_logo_")).toBeTruthy();
    // expect to see text
    expect(getByText("Let's meet you")).toBeTruthy();
    // expect to see an input with placeholder
    expect(getByTestId("name_input").getAttribute("type")).toBe("text");
    // expect to see a button with text - next
    expect(getByText("Next")).toBeTruthy();
    // expect button to be disabled on initial render
    expect(
      getByText("Next").parentElement.matches(
        ".cursor-not-allowed.bg-opacity-50"
      )
    ).toBeTruthy();
    // nothing happens on click of disabled button
    await fireEvent.click(getByText("Next").parentElement);
    expect(
      getByText("Next").parentElement.matches(
        ".cursor-not-allowed.bg-opacity-50"
      )
    ).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  // test that name is updated when input is changed
  test("name is updated when input is changed", async () => {
    const route = "java";
    const setName = () => {};
    // jest.spyOn(setName);
    const history = createMemoryHistory();
    const { container, getByText, getByAltText, getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <AppContext.Provider value={{ route, setName }}>
          <Name />
        </AppContext.Provider>
      </Router>
    );

    // fire a change event on input
    await act(async () => {
      await fireEvent.change(getByTestId("name_input"), {
        target: { value: "Job" },
      });
      // expect value to be updated
      expect(getByTestId("name_input").value).toBe("Job");
      expect(
        getByText("Next").parentElement.matches(".cursor-pointer")
      ).toBeTruthy();
      expect(container.firstChild).toMatchSnapshot();

      // navigates to languages assessment page on button click
      await fireEvent.click(getByText("Next").parentElement);
      expect(mockedNavigate).toHaveBeenCalledWith("/assessment/java");
    });
  });
});
