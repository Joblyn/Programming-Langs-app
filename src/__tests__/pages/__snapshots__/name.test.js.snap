// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Name /> name is updated when input is changed 1`] = `
<div
  class="pt-10"
>
  <div
    class="flex items-center"
  >
    <span
      class="align-middle mr-3"
    >
      <a
        href="/"
      >
        <img
          alt="_logo_"
          class="w-10 h-10"
          src="name_icon.svg"
        />
      </a>
    </span>
    <span
      class="uppercase text-green text-lg font-bold "
    >
      Let's meet you
    </span>
  </div>
  <div>
    <h3
      class="font-bold text-2xl md:text-3xl mb-4"
    >
      First things first, who is taking the assessment?
    </h3>
  </div>
  <div>
    <input
      class="w-full lg:max-w-sm bg-white text-2xl placeholder:text-opacity-50 tracking-wider placeholder:tracking-wider placeholder:text-black-primary p-3 rounded-md border border-green text-black-primary my-3 font-bold focus:outline-0 focus:outline-offset-0"
      data-testid="name_input"
      id="name_input"
      name="name_input"
      placeholder="Your name"
      type="text"
    />
    <div
      class="flex"
    >
      <div
        class="w-full lg:max-w-sm text-white bg-green active:scale-95 uppercase cursor-pointer hover:opacity-80  p-3 text-2xl flex justify-center items-center rounded-md my-3 btn"
      >
        <span
          class="px-[7px] font-bold tracking-wider"
        >
          Next
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Name /> renders correctly 1`] = `
<div
  class="pt-10"
>
  <div
    class="flex items-center"
  >
    <span
      class="align-middle mr-3"
    >
      <a
        href="/"
      >
        <img
          alt="_logo_"
          class="w-10 h-10"
          src="name_icon.svg"
        />
      </a>
    </span>
    <span
      class="uppercase text-green text-lg font-bold "
    >
      Let's meet you
    </span>
  </div>
  <div>
    <h3
      class="font-bold text-2xl md:text-3xl mb-4"
    >
      First things first, who is taking the assessment?
    </h3>
  </div>
  <div>
    <input
      class="w-full lg:max-w-sm bg-white text-2xl placeholder:text-opacity-50 tracking-wider placeholder:tracking-wider placeholder:text-black-primary p-3 rounded-md border border-green text-black-primary my-3 font-bold focus:outline-0 focus:outline-offset-0"
      data-testid="name_input"
      id="name_input"
      name="name_input"
      placeholder="Your name"
      type="text"
    />
    <div
      class="flex"
    >
      <div
        class="w-full lg:max-w-sm text-white bg-green active:scale-95 uppercase cursor-not-allowed bg-opacity-50  p-3 text-2xl flex justify-center items-center rounded-md my-3 btn"
      >
        <span
          class="px-[7px] font-bold tracking-wider"
        >
          Next
        </span>
      </div>
    </div>
  </div>
</div>
`;
